File: intelligent_image_generator.py
#!/usr/bin/env python3
"""
Intelligent Image Generator for SDK3 Homey Apps
Creates professional device images following Johan Bendz design standards
"""

import os
import json
from PIL import Image, ImageDraw
from pathlib import Path

class HomeyImageGenerator:
    def __init__(self, project_path):
        self.project_path = Path(project_path)
        self.assets_path = self.project_path / "assets" / "images"
        self.drivers_path = self.project_path / "drivers"
        
        # SDK3 Homey image specifications
        self.small_size = (75, 75)
        self.large_size = (500, 500)
        
        # Johan Bendz color palette and design standards
        self.color_palette = {
            'sensor': '#4CAF50',      # Green for sensors
            'light': '#FFC107',       # Amber for lights
            'switch': '#2196F3',      # Blue for switches
            'plug': '#FF5722',        # Red-orange for plugs
            'climate': '#9C27B0',     # Purple for climate
            'cover': '#795548',       # Brown for covers/motors
            'detector': '#F44336',    # Red for detectors
            'multisensor': '#00BCD4', # Cyan for multisensors
            'remote': '#607D8B',      # Blue-grey for remotes
            'thermostat': '#E91E63'   # Pink for thermostats
        }
        
        self.device_categories = {
            'air_quality_sensor': 'sensor',
            'co_detector': 'detector',
            'contact_sensor': 'sensor', 
            'curtain_motor': 'cover',
            'dimmer_switch': 'switch',
            'energy_plug': 'plug',
            'light_switch': 'switch',
            'motion_sensor': 'sensor',
            'multisensor': 'multisensor',
            'presence_sensor': 'sensor',
            'rgb_light': 'light',
            'scene_remote_2gang': 'remote',
            'scene_remote_4gang': 'remote', 
            'scene_switch': 'switch',
            'smart_light': 'light',
            'smart_plug': 'plug',
            'smoke_detector': 'detector',
            'temperature_humidity_sensor': 'sensor',
            'thermostat': 'thermostat',
            'water_leak_detector': 'detector'
        }
        
    def create_device_icon(self, device_type, size, style='modern'):
        """Create a professional device icon following SDK3 standards"""
        category = self.device_categories.get(device_type, 'sensor')
        color = self.color_palette[category]
        
        # Create image with transparency
        image = Image.new('RGBA', size, (0, 0, 0, 0))
        draw = ImageDraw.Draw(image)
        
        # Calculate dimensions
        center_x, center_y = size[0] // 2, size[1] // 2
        radius = min(size) // 3
        
        # Create background circle with gradient effect
        self._draw_gradient_circle(draw, center_x, center_y, radius, color, size)
        
        # Add device-specific icon
        self._draw_device_icon(draw, device_type, center_x, center_y, radius, size)
        
        return image
    
    def _draw_gradient_circle(self, draw, center_x, center_y, radius, base_color, size):
        """Draw a gradient circle background"""
        # Convert hex color to RGB
        if base_color.startswith('#'):
            base_color = base_color[1:]
        r = int(base_color[:2], 16)
        g = int(base_color[2:4], 16) 
        b = int(base_color[4:6], 16)
        
        # Create gradient effect
        for i in range(radius):
            alpha = int(255 * (1 - i / radius) * 0.8)  # Fade from center
            circle_color = (r, g, b, alpha)
            draw.ellipse([
                center_x - radius + i, center_y - radius + i,
                center_x + radius - i, center_y + radius - i
            ], fill=circle_color)
    
    def _draw_device_icon(self, draw, device_type, center_x, center_y, radius, size):
        """Draw device-specific icon symbols"""
        icon_size = radius // 2
        white = (255, 255, 255, 255)
        
        if 'sensor' in device_type:
            # Draw sensor symbol (circle with waves)
            draw.ellipse([
                center_x - icon_size//3, center_y - icon_size//3,
                center_x + icon_size//3, center_y + icon_size//3
            ], fill=white)
            
            # Add wave lines
            for i in range(3):
                offset = (i + 1) * icon_size // 4
                draw.ellipse([
                    center_x - offset, center_y - offset,
                    center_x + offset, center_y + offset
                ], outline=white, width=2)
        
        elif 'light' in device_type:
            # Draw light bulb symbol
            bulb_radius = icon_size // 2
            draw.ellipse([
                center_x - bulb_radius, center_y - bulb_radius - icon_size//4,
                center_x + bulb_radius, center_y + bulb_radius - icon_size//4
            ], fill=white)
            
            # Add light rays
            for angle in [0, 45, 90, 135, 180, 225, 270, 315]:
                import math
                ray_x = center_x + int((bulb_radius + icon_size//3) * math.cos(math.radians(angle)))
                ray_y = center_y + int((bulb_radius + icon_size//3) * math.sin(math.radians(angle))) - icon_size//4
                draw.line([center_x, center_y - icon_size//4, ray_x, ray_y], fill=white, width=2)
        
        elif 'switch' in device_type:
            # Draw switch symbol
            switch_width = icon_size
            switch_height = icon_size // 3
            draw.rectangle([
                center_x - switch_width//2, center_y - switch_height//2,
                center_x + switch_width//2, center_y + switch_height//2
            ], fill=white)
            
            # Add switch toggle
            toggle_size = switch_height // 2
            draw.ellipse([
                center_x - toggle_size, center_y - toggle_size,
                center_x + toggle_size, center_y + toggle_size
            ], fill=(100, 100, 100, 255))
        
        elif 'plug' in device_type:
            # Draw power plug symbol
            plug_width = icon_size
            plug_height = icon_size // 2
            draw.rectangle([
                center_x - plug_width//2, center_y - plug_height//4,
                center_x + plug_width//2, center_y + plug_height//4
            ], fill=white)
            
            # Add prongs
            prong_width = plug_width // 6
            prong_height = plug_height // 2
            draw.rectangle([
                center_x - plug_width//3, center_y - plug_height//2,
                center_x - plug_width//3 + prong_width, center_y - plug_height//4
            ], fill=white)
            draw.rectangle([
                center_x + plug_width//3 - prong_width, center_y - plug_height//2,
                center_x + plug_width//3, center_y - plug_height//4
            ], fill=white)
        
        elif 'detector' in device_type:
            # Draw detector symbol (warning triangle or circle)
            if 'smoke' in device_type or 'co' in device_type:
                # Draw warning triangle
                import math
                triangle_size = icon_size
                points = [
                    (center_x, center_y - triangle_size//2),
                    (center_x - triangle_size//2, center_y + triangle_size//2),
                    (center_x + triangle_size//2, center_y + triangle_size//2)
                ]
                draw.polygon(points, fill=white)
                
                # Add exclamation mark
                draw.line([center_x, center_y - triangle_size//4, center_x, center_y + triangle_size//8], 
                         fill=(255, 0, 0, 255), width=3)
                draw.ellipse([
                    center_x - 2, center_y + triangle_size//4 - 2,
                    center_x + 2, center_y + triangle_size//4 + 2
                ], fill=(255, 0, 0, 255))
        
        elif 'remote' in device_type:
            # Draw remote control symbol
            remote_width = icon_size // 2
            remote_height = icon_size
            draw.rectangle([
                center_x - remote_width//2, center_y - remote_height//2,
                center_x + remote_width//2, center_y + remote_height//2
            ], fill=white)
            
            # Add buttons
            button_size = remote_width // 4
            for i in range(3):
                button_y = center_y - remote_height//3 + i * remote_height//4
                draw.ellipse([
                    center_x - button_size, button_y - button_size,
                    center_x + button_size, button_y + button_size
                ], fill=(100, 100, 100, 255))
    
    def generate_all_driver_images(self):
        """Generate images for all drivers in the project"""
        results = {
            'generated': [],
            'errors': [],
            'total_drivers': 0
        }
        
        # Scan all driver directories
        for driver_dir in self.drivers_path.iterdir():
            if driver_dir.is_dir() and not driver_dir.name.startswith('.'):
                results['total_drivers'] += 1
                
                try:
                    # Create assets directory if it doesn't exist
                    assets_dir = driver_dir / "assets" / "images"
                    assets_dir.mkdir(parents=True, exist_ok=True)
                    
                    # Generate small image (75x75)
                    small_image = self.create_device_icon(driver_dir.name, self.small_size)
                    small_path = assets_dir / "small.png"
                    small_image.save(small_path, "PNG")
                    
                    # Generate large image (500x500) 
                    large_image = self.create_device_icon(driver_dir.name, self.large_size)
                    large_path = assets_dir / "large.png"
                    large_image.save(large_path, "PNG")
                    
                    results['generated'].append({
                        'driver': driver_dir.name,
                        'small_image': str(small_path),
                        'large_image': str(large_path)
                    })
                    
                    print(f"[OK] Generated images for {driver_dir.name}")
                    
                except Exception as e:
                    error_msg = f"Failed to generate images for {driver_dir.name}: {str(e)}"
                    results['errors'].append(error_msg)
                    print(f"[ERROR] {error_msg}")
        
        # Save generation report
        report_path = self.project_path / "image_generation_report.json"
        with open(report_path, 'w') as f:
            json.dump(results, f, indent=2)
        
        print(f"\nImage Generation Complete!")
        print(f"Generated images for {len(results['generated'])} drivers")
        print(f"Errors: {len(results['errors'])}")
        print(f"Report saved to: {report_path}")
        
        return results

def main():
    project_path = r"c:\Users\HP\Desktop\tuya_repair"
    generator = HomeyImageGenerator(project_path)
    results = generator.generate_all_driver_images()
    
    if results['errors']:
        print("\nErrors encountered:")
        for error in results['errors']:
            print(f"  - {error}")

if __name__ == "__main__":
    main()

File: ultimate_system/docs_moved/scripts/organization/intelligent-image-generator.js
#!/usr/bin/env node

/**
 * Intelligent Image Generator for Ultimate Zigbee Hub
 * Following Johan Benz design standards and Homey SDK3 guidelines
 * Generates professional device images with correct dimensions
 */

const fs = require('fs-extra');
const path = require('path');
const { createCanvas, loadImage } = require('canvas');

class IntelligentImageGenerator {
    constructor() {
        this.projectRoot = process.cwd();
        
        // SDK3 Image Dimensions
        this.dimensions = {
            app: {
                small: { width: 250, height: 175 },
                large: { width: 500, height: 350 },
                xlarge: { width: 1000, height: 700 }
            },
            driver: {
                small: { width: 75, height: 75 },
                large: { width: 500, height: 500 },
                xlarge: { width: 1000, height: 1000 }
            }
        };
        
        // Johan Benz Color Palette by Device Category
        this.colorPalette = {
            sensors: {
                primary: '#2196F3',
                secondary: '#03A9F4',
                gradient: ['#1976D2', '#42A5F5']
            },
            lights: {
                primary: '#FFA500',
                secondary: '#FFD700',
                gradient: ['#FF8F00', '#FFC107']
            },
            switches: {
                primary: '#4CAF50',
                secondary: '#8BC34A',
                gradient: ['#388E3C', '#66BB6A']
            },
            plugs: {
                primary: '#9C27B0',
                secondary: '#673AB7',
                gradient: ['#7B1FA2', '#AB47BC']
            },
            safety: {
                primary: '#F44336',
                secondary: '#E91E63',
                gradient: ['#D32F2F', '#EF5350']
            },
            climate: {
                primary: '#FF9800',
                secondary: '#FF5722',
                gradient: ['#F57C00', '#FF7043']
            },
            covers: {
                primary: '#607D8B',
                secondary: '#455A64',
                gradient: ['#455A64', '#78909C']
            },
            specialty: {
                primary: '#795548',
                secondary: '#8D6E63',
                gradient: ['#5D4037', '#A1887F']
            }
        };
        
        // Device icons (simple geometric representations)
        this.deviceIcons = {
            motion_sensor: this.drawMotionSensorIcon,
            contact_sensor: this.drawContactSensorIcon,
            temperature_sensor: this.drawTemperatureSensorIcon,
            smart_light: this.drawLightIcon,
            smart_plug: this.drawPlugIcon,
            switch: this.drawSwitchIcon,
            smoke_detector: this.drawSmokeDetectorIcon,
            curtain_motor: this.drawCurtainIcon,
            thermostat: this.drawThermostatIcon,
            button: this.drawButtonIcon,
            dimmer: this.drawDimmerIcon,
            rgb_light: this.drawRGBLightIcon,
            presence_sensor: this.drawPresenceSensorIcon,
            water_leak: this.drawWaterLeakIcon,
            air_quality: this.drawAirQualityIcon
        };
    }

    async generateAllImages() {
        console.log('🎨 Starting intelligent image generation...');
        
        try {
            // Generate app images
            await this.generateAppImages();
            
            // Generate driver images
            await this.generateDriverImages();
            
            // Validate all generated images
            await this.validateImages();
            
            console.log('✅ All images generated successfully!');
            
        } catch (error) {
            console.error('❌ Error generating images:', error);
            throw error;
        }
    }

    async generateAppImages() {
        console.log('📱 Generating app images...');
        
        const appImagesDir = path.join(this.projectRoot, 'assets/images');
        await fs.ensureDir(appImagesDir);
        
        // Generate main app images with Ultimate Zigbee Hub branding
        for (const [size, dimensions] of Object.entries(this.dimensions.app)) {
            const canvas = createCanvas(dimensions.width, dimensions.height);
            const ctx = canvas.getContext('2d');
            
            // Create gradient background
            const gradient = ctx.createLinearGradient(0, 0, dimensions.width, dimensions.height);
            gradient.addColorStop(0, '#1976D2');
            gradient.addColorStop(1, '#42A5F5');
            
            ctx.fillStyle = gradient;
            ctx.fillRect(0, 0, dimensions.width, dimensions.height);
            
            // Add Zigbee symbol and hub representation
            this.drawZigbeeHubIcon(ctx, dimensions);
            
            // Add text for larger sizes
            if (size !== 'small') {
                this.drawAppTitle(ctx, dimensions, 'Ultimate Zigbee Hub');
            }
            
            const buffer = canvas.toBuffer('image/png');
            await fs.writeFile(path.join(appImagesDir, `${size}.png`), buffer);
            console.log(`📷 Generated app image: ${size}.png (${dimensions.width}x${dimensions.height})`);
        }
    }

    async generateDriverImages() {
        console.log('🔧 Generating driver images...');
        
        const driversDir = path.join(this.projectRoot, 'drivers');
        if (!await fs.pathExists(driversDir)) return;
        
        const categories = await fs.readdir(driversDir);
        
        for (const category of categories) {
            const categoryPath = path.join(driversDir, category);
            const stat = await fs.stat(categoryPath);
            
            if (stat.isDirectory()) {
                const drivers = await fs.readdir(categoryPath);
                
                for (const driver of drivers) {
                    const driverPath = path.join(categoryPath, driver);
                    const driverStat = await fs.stat(driverPath);
                    
                    if (driverStat.isDirectory()) {
                        await this.generateDriverImageSet(category, driver, driverPath);
                    }
                }
            }
        }
    }

    async generateDriverImageSet(category, driverName, driverPath) {
        const assetsDir = path.join(driverPath, 'assets');
        await fs.ensureDir(assetsDir);
        
        const colors = this.colorPalette[category] || this.colorPalette.specialty;
        const deviceType = this.detectDeviceType(driverName);
        
        // Generate all required driver image sizes
        for (const [size, dimensions] of Object.entries(this.dimensions.driver)) {
            const canvas = createCanvas(dimensions.width, dimensions.height);
            const ctx = canvas.getContext('2d');
            
            // Create professional gradient background
            const gradient = ctx.createRadialGradient(
                dimensions.width / 2, dimensions.height / 2, 0,
                dimensions.width / 2, dimensions.height / 2, dimensions.width / 2
            );
            gradient.addColorStop(0, '#FFFFFF');
            gradient.addColorStop(0.7, '#F8F9FA');
            gradient.addColorStop(1, '#E9ECEF');
            
            ctx.fillStyle = gradient;
            ctx.fillRect(0, 0, dimensions.width, dimensions.height);
            
            // Draw device icon
            await this.drawDeviceIcon(ctx, dimensions, deviceType, colors);
            
            // Add subtle shadow/depth effect
            this.addShadowEffect(ctx, dimensions);
            
            const buffer = canvas.toBuffer('image/png');
            const imageName = size === 'xlarge' ? 'large.png' : `${size}.png`;
            await fs.writeFile(path.join(assetsDir, imageName), buffer);
            console.log(`🖼️  Generated ${category}/${driverName}: ${imageName}`);
        }
    }

    detectDeviceType(driverName) {
        const name = driverName.toLowerCase();
        
        if (name.includes('motion') || name.includes('pir')) return 'motion_sensor';
        if (name.includes('contact') || name.includes('door') || name.includes('window')) return 'contact_sensor';
        if (name.includes('temperature') || name.includes('humidity')) return 'temperature_sensor';
        if (name.includes('light') || name.includes('bulb') || name.includes('lamp')) return 'smart_light';
        if (name.includes('plug') || name.includes('socket')) return 'smart_plug';
        if (name.includes('switch') && !name.includes('dimmer')) return 'switch';
        if (name.includes('smoke') || name.includes('detector')) return 'smoke_detector';
        if (name.includes('curtain') || name.includes('motor') || name.includes('blind')) return 'curtain_motor';
        if (name.includes('thermostat') || name.includes('radiator')) return 'thermostat';
        if (name.includes('button') || name.includes('scene')) return 'button';
        if (name.includes('dimmer')) return 'dimmer';
        if (name.includes('rgb') || name.includes('color')) return 'rgb_light';
        if (name.includes('presence') || name.includes('radar')) return 'presence_sensor';
        if (name.includes('water') || name.includes('leak')) return 'water_leak';
        if (name.includes('air') || name.includes('quality')) return 'air_quality';
        
        return 'switch'; // Default fallback
    }

    async drawDeviceIcon(ctx, dimensions, deviceType, colors) {
        const iconFunction = this.deviceIcons[deviceType] || this.deviceIcons.switch;
        await iconFunction.call(this, ctx, dimensions, colors);
    }

    drawMotionSensorIcon(ctx, dimensions, colors) {
        const centerX = dimensions.width / 2;
        const centerY = dimensions.height / 2;
        const radius = Math.min(dimensions.width, dimensions.height) * 0.25;
        
        // Main sensor body
        ctx.fillStyle = colors.primary;
        ctx.beginPath();
        ctx.roundRect(centerX - radius, centerY - radius, radius * 2, radius * 2, radius * 0.2);
        ctx.fill();
        
        // Detection waves
        ctx.strokeStyle = colors.secondary;
        ctx.lineWidth = dimensions.width / 50;
        for (let i = 1; i <= 3; i++) {
            ctx.beginPath();
            ctx.arc(centerX, centerY, radius * (1 + i * 0.3), 0, Math.PI * 2);
            ctx.stroke();
        }
        
        // Lens
        ctx.fillStyle = '#FFFFFF';
        ctx.beginPath();
        ctx.arc(centerX, centerY, radius * 0.4, 0, Math.PI * 2);
        ctx.fill();
        
        ctx.fillStyle = colors.secondary;
        ctx.beginPath();
        ctx.arc(centerX, centerY, radius * 0.2, 0, Math.PI * 2);
        ctx.fill();
    }

    drawContactSensorIcon(ctx, dimensions, colors) {
        const centerX = dimensions.width / 2;
        const centerY = dimensions.height / 2;
        const width = Math.min(dimensions.width, dimensions.height) * 0.3;
        const height = width * 1.5;
        
        // Sensor base
        ctx.fillStyle = colors.primary;
        ctx.beginPath();
        ctx.roundRect(centerX - width/2, centerY - height/2, width, height, width * 0.1);
        ctx.fill();
        
        // Magnetic part
        ctx.fillStyle = colors.secondary;
        ctx.beginPath();
        ctx.roundRect(centerX + width/2 + 5, centerY - height/4, width/2, height/2, width * 0.05);
        ctx.fill();
        
        // Connection lines
        ctx.strokeStyle = colors.gradient[0];
        ctx.lineWidth = 2;
        ctx.setLineDash([5, 5]);
        ctx.beginPath();
        ctx.moveTo(centerX + width/2, centerY - height/4);
        ctx.lineTo(centerX + width/2 + 5, centerY - height/4);
        ctx.moveTo(centerX + width/2, centerY + height/4);
        ctx.lineTo(centerX + width/2 + 5, centerY + height/4);
        ctx.stroke();
        ctx.setLineDash([]);
    }

    drawLightIcon(ctx, dimensions, colors) {
        const centerX = dimensions.width / 2;
        const centerY = dimensions.height / 2;
        const radius = Math.min(dimensions.width, dimensions.height) * 0.2;
        
        // Bulb shape
        ctx.fillStyle = colors.primary;
        ctx.beginPath();
        ctx.arc(centerX, centerY - radius * 0.2, radius, 0, Math.PI * 2);
        ctx.fill();
        
        // Base/screw
        ctx.fillStyle = colors.secondary;
        ctx.beginPath();
        ctx.rect(centerX - radius * 0.6, centerY + radius * 0.8, radius * 1.2, radius * 0.4);
        ctx.fill();
        
        // Light rays
        ctx.strokeStyle = colors.gradient[1];
        ctx.lineWidth = 3;
        const rayLength = radius * 0.8;
        for (let i = 0; i < 8; i++) {
            const angle = (i * Math.PI * 2) / 8;
            const startX = centerX + Math.cos(angle) * (radius + 10);
            const startY = centerY - radius * 0.2 + Math.sin(angle) * (radius + 10);
            const endX = centerX + Math.cos(angle) * (radius + rayLength);
            const endY = centerY - radius * 0.2 + Math.sin(angle) * (radius + rayLength);
            
            ctx.beginPath();
            ctx.moveTo(startX, startY);
            ctx.lineTo(endX, endY);
            ctx.stroke();
        }
    }

    drawPlugIcon(ctx, dimensions, colors) {
        const centerX = dimensions.width / 2;
        const centerY = dimensions.height / 2;
        const size = Math.min(dimensions.width, dimensions.height) * 0.3;
        
        // Outlet face
        ctx.fillStyle = colors.primary;
        ctx.beginPath();
        ctx.roundRect(centerX - size, centerY - size, size * 2, size * 2, size * 0.15);
        ctx.fill();
        
        // Outlet holes
        ctx.fillStyle = '#FFFFFF';
        ctx.beginPath();
        ctx.arc(centerX - size * 0.3, centerY - size * 0.2, size * 0.15, 0, Math.PI * 2);
        ctx.arc(centerX + size * 0.3, centerY - size * 0.2, size * 0.15, 0, Math.PI * 2);
        ctx.rect(centerX - size * 0.1, centerY + size * 0.1, size * 0.2, size * 0.3);
        ctx.fill();
        
        // Power indicator
        ctx.fillStyle = colors.secondary;
        ctx.beginPath();
        ctx.arc(centerX, centerY - size * 0.7, size * 0.1, 0, Math.PI * 2);
        ctx.fill();
    }

    drawZigbeeHubIcon(ctx, dimensions) {
        const centerX = dimensions.width / 2;
        const centerY = dimensions.height / 2;
        const size = Math.min(dimensions.width, dimensions.height) * 0.4;
        
        // Hub center
        ctx.fillStyle = '#FFFFFF';
        ctx.beginPath();
        ctx.arc(centerX, centerY, size * 0.2, 0, Math.PI * 2);
        ctx.fill();
        
        // Connection nodes
        const nodePositions = [
            { x: centerX - size * 0.4, y: centerY - size * 0.3 },
            { x: centerX + size * 0.4, y: centerY - size * 0.3 },
            { x: centerX - size * 0.4, y: centerY + size * 0.3 },
            { x: centerX + size * 0.4, y: centerY + size * 0.3 },
            { x: centerX, y: centerY - size * 0.5 },
            { x: centerX, y: centerY + size * 0.5 }
        ];
        
        // Draw connections
        ctx.strokeStyle = 'rgba(255,255,255,0.8)';
        ctx.lineWidth = 2;
        nodePositions.forEach(node => {
            ctx.beginPath();
            ctx.moveTo(centerX, centerY);
            ctx.lineTo(node.x, node.y);
            ctx.stroke();
        });
        
        // Draw nodes
        ctx.fillStyle = '#FFFFFF';
        nodePositions.forEach(node => {
            ctx.beginPath();
            ctx.arc(node.x, node.y, size * 0.08, 0, Math.PI * 2);
            ctx.fill();
        });
    }

    drawAppTitle(ctx, dimensions, title) {
        ctx.fillStyle = '#FFFFFF';
        ctx.font = `bold ${dimensions.height * 0.08}px Arial`;
        ctx.textAlign = 'center';
        ctx.textBaseline = 'bottom';
        ctx.fillText(title, dimensions.width / 2, dimensions.height - 20);
    }

    addShadowEffect(ctx, dimensions) {
        // Add subtle inner shadow for depth
        const gradient = ctx.createRadialGradient(
            dimensions.width / 2, dimensions.height / 2, 0,
            dimensions.width / 2, dimensions.height / 2, dimensions.width / 2
        );
        gradient.addColorStop(0, 'transparent');
        gradient.addColorStop(0.8, 'transparent');
        gradient.addColorStop(1, 'rgba(0,0,0,0.1)');
        
        ctx.fillStyle = gradient;
        ctx.fillRect(0, 0, dimensions.width, dimensions.height);
    }

    async validateImages() {
        console.log('✅ Validating generated images...');
        
        const errors = [];
        
        // Validate app images
        const appImagesDir = path.join(this.projectRoot, 'assets/images');
        for (const [size, expectedDimensions] of Object.entries(this.dimensions.app)) {
            const imagePath = path.join(appImagesDir, `${size}.png`);
            if (!await fs.pathExists(imagePath)) {
                errors.push(`Missing app image: ${size}.png`);
            }
        }
        
        // Validate driver images
        const driversDir = path.join(this.projectRoot, 'drivers');
        if (await fs.pathExists(driversDir)) {
            // Check each driver has required images
            // This would be expanded based on actual driver structure
        }
        
        if (errors.length > 0) {
            console.error('❌ Image validation errors:', errors);
            throw new Error('Image validation failed');
        }
        
        console.log('✅ All images validated successfully');
    }

    // Additional icon drawing methods would be implemented here
    drawTemperatureSensorIcon(ctx, dimensions, colors) { /* Implementation */ }
    drawSwitchIcon(ctx, dimensions, colors) { /* Implementation */ }
    drawSmokeDetectorIcon(ctx, dimensions, colors) { /* Implementation */ }
    drawCurtainIcon(ctx, dimensions, colors) { /* Implementation */ }
    drawThermostatIcon(ctx, dimensions, colors) { /* Implementation */ }
    drawButtonIcon(ctx, dimensions, colors) { /* Implementation */ }
    drawDimmerIcon(ctx, dimensions, colors) { /* Implementation */ }
    drawRGBLightIcon(ctx, dimensions, colors) { /* Implementation */ }
    drawPresenceSensorIcon(ctx, dimensions, colors) { /* Implementation */ }
    drawWaterLeakIcon(ctx, dimensions, colors) { /* Implementation */ }
    drawAirQualityIcon(ctx, dimensions, colors) { /* Implementation */ }
}

// Execute if run directly
if (require.main === module) {
    const generator = new IntelligentImageGenerator();
    generator.generateAllImages().catch(console.error);
}

module.exports = IntelligentImageGenerator;
